st2actions/st2actions/runners/ssh/paramiko_ssh.py

```diff
@@ -514,6 +524,51 @@ def _get_pkey_object(self, key_material):
 
         raise paramiko.ssh_exception.SSHException(msg)
 
+    def _bastion_connect(self):
+        """
+        Connect to the bastion node over SSH.
+
+        :return: True if the connection has been successfully established,
+                 False otherwise.
+        :rtype: ``bool``
+        """
+        bastion_conninfo = {'hostname': self.bastion_host,
+                            'port': self.port,
+                            'username': self.username,
+                            'allow_agent': False,
+                            'look_for_keys': False,
+                            'timeout': self.timeout}
+
+        if self.password:
+            bastion_conninfo['password'] = self.password
+
+        if self.key_files:
+            bastion_conninfo['key_filename'] = self.key_files
+
+        if self.key_material:
+            bastion_conninfo['pkey'] = self._get_pkey_object(key_material=self.key_material)
+
+        if not self.password and not (self.key_files or self.key_material):
+            bastion_conninfo['allow_agent'] = True
+            bastion_conninfo['look_for_keys'] = True
+
+        bastion_extra = {'_hostname': self.bastion_host,
+                         '_port': self.port,
+                         '_username': self.username,
+                         '_timeout': self.timeout}
+
+        self.logger.debug('Connecting to bastion host', extra=bastion_extra)
+
+        self.bastion_client.connect(**bastion_conninfo)
+
+        transport = self.bastion_client.get_transport()
+        real_addr = (self.hostname, self.port)
+        # no attempt is made to bind local_addr, so something very "wrong" is used
+        local_addr = ('256.256.256.256', 65536)
```

Comment:
Fabric will deprecated soon. We still have it to make sure we didn't miss any feature in fabric that hasn't been ported over to.